### PR TITLE
Procs Initial Refactor

### PR DESCRIPTION
**Pull request by scopatz**
This is an initial refactor of the subproc support.  This is meant as an incremental organizational step to being able to do deeper refactors going forward. This doesn't change any xonsh behavior. It just moves code from `built_ins.py` and `proc.py` into a new `procs` subpackage.  It starts to separate things into platform-specific implementations, but does not fully do that yet (in the spirit of not changing actual functionality).

I also took this opportunity to do some generic code cleanup, such as changing things to f-strings and other minor improvements.

### CODE REVIEW

**Comment by scopatz** [comment]
@gforsyth - nort sure why doctr is failing here when I update the amalgamate hooks.  Any ideas?

**Comment by gforsyth** [comment]
Will try to dig in later -- something is mutating `xonsh/__init__.py` in a way that `doctr` is unhappy about.  But I'm confused as to why the amalgamate hooks made this happen but that it hasn't been a problem in the past

**Comment by gforsyth** [comment]
Ok, so _something_ (don't know what) is messing with the `restore_version` call in `setup.py` that is supposed to undo the `devN` suffix that we add to `__init__.py` when we install from source.  Because of this, we end up with unstaged changes and `doctr` is unhappy.  

I have no idea why this just started happening.  Working on ugly workarounds.

**Comment by gforsyth** [comment]
Ok, Travis is passing now.  

I am still very confused about why this just started happening, and also why my fix works.  
For some reason, if we `python setup.py install`, then our `setup.py` file doesn't get run correctly and we leave a dirty `__init__.py`.  But if we `pip install`, then it gets cleaned up?

Could be that we're doing a runaround on the whole thing since `pip` build isolation stuff might be just transferring the issue into some random temp directory, but I really don't know.

**Comment by scopatz** [comment]
> Could be that we're doing a runaround on the whole thing since pip build isolation stuff might be just transferring the issue into some random temp directory, but I really don't know.

Yeah, I think this is exactly what is happening. Thanks for the fix, though!

**Comment by scopatz** [comment]
I think this is ready for review again @xonsh/xore

**Review by gforsyth — APPROVED** [review]
This looks good to me @scopatz !  Just the one typo in the docs that needs fixing.

File: docs/api/procs/pipelines.rst [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.. _xonsh_procs_piplines:
```
[the remark this answers is not in the document]
Comment (scopatz) [reply docs/api/procs/pipelines.rst:1]: Amazing catch!

**Comment by gforsyth** [comment]
Thanks, @scopatz !

**Comment by scopatz** [comment]
Thanks @gforsyth!

**Pull request merged**